src/literacy.rs

```diff
@@ -0,0 +1,263 @@
+// Bitcoin Hashes Library
+// Written in 2021 by
+//   The rust-bitcoin developers.
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! # Literacy traits
+//!
+//! literacy: "the ability to [Read] and [Write]."
+//!
+//! * With `std` enabled, traits are automatically implemented for `std::io::{Read, Write}`
+//! * Without `std` but using feature `use-core2`, they are implemented for `core2::io::{Read, Write}`
+//! * Without neither `std` and nor `core2` default implementation `impl Read for &[u8]`
+//!   and `impl Write for ::alloc::vec::Vec<u8>` are provided
+//!
+
+/// The Read trait allows for reading bytes from a source.
+pub trait Read {
+    /// The error type returned in Result
+    type Error;
+    /// The type to implement limited reads
+    type Take;
```

Comment:
I would add a trait bound to literacy::Read here, it might help later on when working with generics

Reply:
Alternatively you could also add another trait specifically for the Take type that "extends" `Read` and forces a specific constructor-like method. With that you can then provide a default implementation for `take()` like std does, because you can always construct an object of type `Self::Take` using the constructor defined by its trait.

Reply:
added trait bound in 6b7b28fc3536fda034d70d39a5d1bded78f07670